finetune/adapter.py

```diff
@@ -149,34 +150,36 @@ def fit(
     model: GPT,
     optimizer: torch.optim.Optimizer,
     scheduler: torch.optim.lr_scheduler,
-    train_data: List[Dict],
-    val_data: List[Dict],
+    train_dataloader: DataLoader,
+    val_dataloader: DataLoader,
     devices: int,
     io: IOArgs,
     train: TrainArgs,
     eval: EvalArgs,
 ) -> None:
     tokenizer = Tokenizer(io.checkpoint_dir)
-    longest_seq_length, longest_seq_ix = get_longest_seq_length(train_data)
+    longest_seq_length, longest_seq_ix = get_longest_seq_length(train_dataloader.dataset)
     model.max_seq_length = min(longest_seq_length, train.max_seq_length or float("inf"))
     fabric.print(
         f"The longest sequence length in the train data is {longest_seq_length}, the model's maximum sequence length is"
         f" {model.max_seq_length} and context length is {model.config.block_size}"
     )
 
-    validate(fabric, model, val_data, tokenizer, dataclasses.replace(eval, max_iters=2), train)  # sanity check
+    validate(fabric, model, val_dataloader, tokenizer, dataclasses.replace(eval, max_iters=2))  # sanity check
 
+    train_iterator = CycleIterator(train_dataloader)
     throughput = ThroughputMonitor(fabric, window_size=50)
     step_count = 0
+    iter_num = 0
     total_lengths = 0
     total_t0 = time.perf_counter()
 
-    for iter_num in range(1, train.max_iters(devices) + 1):
+    while step_count < (train.max_steps or float("inf")) and train_iterator.epoch < train.epochs:
```

Comment:
Can you `or train.max_steps` outside the loop?

Reply:
Actually, I don't know what you mean. Can you clarify

Reply:
I didn't write what I was thinking, edited

Reply:
Like `train.max_steps = train.max_steps or float("inf")` for example?

Reply:
Yes but do not modify the config, we should treat it as immutable. Save it into a variable